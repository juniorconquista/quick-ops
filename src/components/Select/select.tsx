import React, { memo } from 'react';
import ReactSelect from 'react-select';
import makeAnimated from 'react-select/animated';
import { SelectComponents } from 'react-select/src/components';
import { Box, Label, P } from 'quarks-ui';
import styled from 'styled-components';
import { variant, VariantArgs } from 'styled-system';

import { SelectProps, SelectStyledProps } from './select.types';

/* TODO This component was created in the local repository as it has not yet been 
implemented in lib Quarks UI, however in the future it can be changed to be used directly from lib. */

const variants = variant({
    prop: 'styling',
    variants: {
        base: {
            width: '100%',
            label: {
                m: '12px 0px',
                color: 'text.contrast',
                fontWeight: 'bold',
                display: 'block',
            },
            '.select__input': {
                color: '#6b7480',
                fontFamily: 'Nunito Sans, sans-serif',
                fontWeight: 'normal',
                '& > div': {
                    fontFamily: 'Nunito Sans, sans-serif !important',
                    fontWeight: 'normal !important',
                },
            },
            '.select__control': {
                border: '1px solid',
                borderRadius: '6px',
                bg: 'white.default',
                borderColor: '#dfe2e6',
                fontFamily: 'Nunito Sans, sans-serif',
                height: '36px',
                transition: 'inherit',
                '&:hover': {
                    boxShadow: '0 0 2px 3px #6b748090',
                    outline: 'none',
                    border: '1px solid',
                    borderColor: '#dfe2e6',
                },
            },
            '.select__placeholder': {
                color: '#6b7480',
                fontFamily: 'Nunito Sans, sans-serif',
            },
            '.select__single-value': {
                color: '#6b7480',
                fontFamily: 'Nunito Sans, sans-serif',
            },
            '.select__control--is-focused': {
                boxShadow: '0 0 2px 3px #6b748090',
                borderColor: '#dfe2e6',
            },
            '.select__menu': {
                bg: 'white.default',
                border: '1px solid',
                borderColor: '#dfe2e6',
                marginTop: '4px',
                overflow: 'hidden',
            },
            '.select__menu-list': {
                color: '#6b7480',
                fontFamily: 'Nunito Sans, sans-serif',
                paddingTop: '0rem',
                paddingBottom: '0rem',
            },
            '.select__option': {
                transition: 'all 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
                lineHeight: 'normal',
                '&:hover': {
                    bg: 'white.contrast',
                },
            },
            '.select__option--is-focused': {
                bg: 'white.contrast',
            },
            '.select__option--is-selected': {
                bg: 'primary.default',
            },
        },
        get success() {
            const base: { [key: string]: any } = this.base;
            return {
                ...base,
                '.select__input': {
                    ...base['.select__input'],
                    color: 'success.default',
                },
                '.select__control': {
                    ...base['.select__control'],
                    borderColor: 'success.default',
                    '&:hover': {
                        ...base['.select__control']['&:hover'],
                        borderColor: 'success.default',
                        boxShadow: (props: {
                            colors: { success: { default: string } };
                        }) => `0 0 2px 3px ${props.colors.success.default}90`,
                    },
                },
                '.select__control--is-focused': {
                    ...base['.select__control--is-focused'],
                    borderColor: 'success.default',
                    boxShadow: (props: {
                        colors: { success: { default: string } };
                    }) => `0 0 2px 3px ${props.colors.success.default}90`,
                },
                '.select__menu': {
                    ...base['.select__menu'],
                    borderColor: 'success.default',
                },
                '.select__menu-list': {
                    ...base['.select__menu-list'],
                    color: 'success.default',
                },
            };
        },
        get danger() {
            const base: { [key: string]: any } = this.base;
            return {
                ...base,
                '.select__input': {
                    ...base['.select__input'],
                    color: 'danger.default',
                },
                '.select__control': {
                    ...base['.select__control'],
                    borderColor: 'danger.default',
                    '&:hover': {
                        ...base['.select__control']['&:hover'],
                        borderColor: 'danger.default',
                        boxShadow: (props: {
                            colors: { danger: { default: string } };
                        }) => `0 0 2px 3px ${props.colors.danger.default}90`,
                    },
                },
                '.select__control--is-focused': {
                    ...base['.select__control--is-focused'],
                    borderColor: 'danger.default',
                    boxShadow: (props: {
                        colors: { danger: { default: string } };
                    }) => `0 0 2px 3px ${props.colors.danger.default}90`,
                },
                '.select__menu': {
                    ...base['.select__menu'],
                    borderColor: 'danger.default',
                },
                '.select__menu-list': {
                    ...base['.select__menu-list'],
                    color: 'danger.default',
                },
            };
        },
    },
} as VariantArgs);

const Style: React.FC<SelectStyledProps> = styled(ReactSelect)<
    SelectStyledProps
>`
    ${variants}
`;

const animatedComponents = makeAnimated() as SelectComponents<{
    label: string;
    value: string;
}>;

const Select: React.FC<SelectProps> = ({
    styling = 'base',
    label,
    error,
    name,
    ...props
}) => (
    <Box styling="column" width="100%" alignItems="flex-start">
        {label && (
            <Label
                m="12px 0px"
                fontWeight="bold"
                color={error ? 'danger.default' : 'text.contrast'}
            >
                {label}
            </Label>
        )}
        <Style
            styling={styling}
            classNamePrefix="select"
            components={animatedComponents}
            id={name}
            {...props}
        />
        {error && (
            <P styling="base" mt="12px" color="danger.default">
                {error}
            </P>
        )}
    </Box>
);

export default memo(Select);
