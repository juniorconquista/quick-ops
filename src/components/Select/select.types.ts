import { Props as ReactSelectProps } from 'react-select';

export type SelectStyledProps = ReactSelectProps & {
    /** Each styling variant has unique properties that are inherited through the type informed */
    styling?: 'base' | 'success' | 'danger';
};

export interface SelectProps extends SelectStyledProps {
    label?: string;
    error?: string | boolean;
}
