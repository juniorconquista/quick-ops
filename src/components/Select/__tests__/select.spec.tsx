import React from 'react';
import { cleanup, render } from '@testing-library/react';

import ThemeProvider from '../../../utils/ThemeProviderTest';

import 'jest-styled-components';
import '@testing-library/jest-dom/extend-expect';

import Select from '../select';

describe('Select component', () => {
    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
    });
    test('basic snapshot', () => {
        const container = render(
            <ThemeProvider>
                <Select styling="base" />
            </ThemeProvider>,
        );
        expect(container.asFragment()).toMatchSnapshot();
    });

    it('should inherit the styling properties of the success variant', () => {
        render(
            <ThemeProvider>
                <Select styling="success" />
            </ThemeProvider>,
        );
    });

    it('should inherit the styling properties of the danger variant', () => {
        render(
            <ThemeProvider>
                <Select styling="danger" />
            </ThemeProvider>,
        );
    });
});
