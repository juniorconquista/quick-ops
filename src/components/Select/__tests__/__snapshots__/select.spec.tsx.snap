// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component basic snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  width: 100%;
}

.c0:focus {
  outline: none;
}

.c1 {
  width: 100%;
}

.c1 label {
  margin: 12px 0px;
  color: #FFFFFF;
  font-weight: bold;
  display: block;
}

.c1 .select__input {
  color: #6b7480;
  font-family: Nunito Sans,sans-serif;
  font-weight: normal;
}

.c1 .select__input > div {
  font-family: Nunito Sans,sans-serif !important;
  font-weight: normal !important;
}

.c1 .select__control {
  border: 1px solid;
  border-radius: 6px;
  background-color: #25262a;
  border-color: #dfe2e6;
  font-family: Nunito Sans,sans-serif;
  height: 36px;
  -webkit-transition: inherit;
  transition: inherit;
}

.c1 .select__control:hover {
  box-shadow: 0 0 2px 3px #6b748090;
  outline: none;
  border: 1px solid;
  border-color: #dfe2e6;
}

.c1 .select__placeholder {
  color: #6b7480;
  font-family: Nunito Sans,sans-serif;
}

.c1 .select__single-value {
  color: #6b7480;
  font-family: Nunito Sans,sans-serif;
}

.c1 .select__control--is-focused {
  box-shadow: 0 0 2px 3px #6b748090;
  border-color: #dfe2e6;
}

.c1 .select__menu {
  background-color: #25262a;
  border: 1px solid;
  border-color: #dfe2e6;
  margin-top: 4px;
  overflow: hidden;
}

.c1 .select__menu-list {
  color: #6b7480;
  font-family: Nunito Sans,sans-serif;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

.c1 .select__option {
  -webkit-transition: all 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: all 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  line-height: normal;
}

.c1 .select__option:hover {
  background-color: #313438;
}

.c1 .select__option--is-focused {
  background-color: #313438;
}

.c1 .select__option--is-selected {
  background-color: #51c1c3;
}

<div
    class="c0"
    width="100%"
  >
    <div
      class="c1 css-2b097c-container"
    >
      <div
        class="select__control css-yk16xz-control"
      >
        <div
          class="select__value-container css-g1d714-ValueContainer"
        >
          <div
            class="select__placeholder css-1wa3eu0-placeholder"
            style="opacity: 1; transition: opacity 1ms;"
          >
            Select...
          </div>
          <div
            class="css-b8ldur-Input"
          >
            <div
              class="select__input"
              style="display: inline-block;"
            >
              <input
                aria-autocomplete="list"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                id="react-select-2-input"
                spellcheck="false"
                style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
              <div
                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
              />
            </div>
          </div>
        </div>
        <div
          class="select__indicators css-1hb7zxy-IndicatorsContainer"
        >
          <span
            class="select__indicator-separator css-1okebmr-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            class="select__indicator select__dropdown-indicator css-tlfecz-indicatorContainer"
          >
            <svg
              aria-hidden="true"
              class="css-6q0nyr-Svg"
              focusable="false"
              height="20"
              viewBox="0 0 20 20"
              width="20"
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
